src/backend/distributed/worker/worker_shard_visibility.c

```diff
@@ -257,72 +266,168 @@ RelationIsAKnownShard(Oid shardRelationId)
 }
 
 
 /*
- * ReplaceTableVisibleFunction replaces all occurrences of
- * pg_catalog.pg_table_visible() to
- * pg_catalog.citus_table_visible() in the given input node.
- *
- * Note that the only difference between the functions is that
- * the latter filters the tables that are known to be shards on
- * Citus MX worker (data) nodes.
- *
- * Note that although the function mutates the input node, we
- * prefer to use query_tree_walker/expression_tree_walker over
- * their mutator equivalents. This is safe because we ensure that
- * the replaced function has the exact same input/output values with
- * its precedent.
+ * ResetHideShardsDecision resets the decision whether to hide shards.
+ */
+void
+ResetHideShardsDecision(void)
+{
+	HideShards = CHECK_APPLICATION_NAME;
+}
+
+
+/*
+ * ShouldHideShardsFromApplication determines whether we should hide shards based on
+ * the current application name.
  */
 static bool
-ReplaceTableVisibleFunctionWalker(Node *inputNode)
+ShouldHideShardsFromApplication(void)
 {
-	if (inputNode == NULL)
+	List *prefixList = NIL;
+
+	/* SplitGUCList scribbles on the input */
+	char *splitCopy = pstrdup(HideShardsFromAppNamePrefixes);
+
+	if (!SplitGUCList(splitCopy, ',', &prefixList))
 	{
+		/* invalid GUC value, ignore */
 		return false;
 	}
 
-	if (IsA(inputNode, FuncExpr))
+	char *appNamePrefix = NULL;
+	foreach_ptr(appNamePrefix, prefixList)
 	{
-		FuncExpr *functionToProcess = (FuncExpr *) inputNode;
-		Oid functionId = functionToProcess->funcid;
-
-		if (functionId == PgTableVisibleFuncId())
+		/* always hide shards when one of the prefixes is * */
+		if (strcmp(appNamePrefix, "*") == 0)
 		{
-			/*
-			 * We simply update the function id of the FuncExpr for
-			 * two reasons: (i) We don't want to interfere with the
-			 * memory contexts so don't want to deal with allocating
-			 * a new functionExpr (ii) We already know that both
-			 * functions have the exact same signature.
-			 */
-			functionToProcess->funcid = CitusTableVisibleFuncId();
+			return true;
+		}
 
-			/* although not very likely, we could have nested calls to pg_table_is_visible */
-			return expression_tree_walker(inputNode, ReplaceTableVisibleFunctionWalker,
-										  NULL);
+		/* compare only the first first <prefixLength> characters */
+		int prefixLength = strlen(appNamePrefix);
+		if (strncmp(application_name, appNamePrefix, prefixLength) == 0)
+		{
+			return true;
 		}
 	}
-	else if (IsA(inputNode, Query))
+
+	return false;
+}
+
+
+/*
+ * FilterShardsFromPgclass adds a NOT relation_is_a_known_shard(oid) filter
+ * to the security quals of pg_class RTEs.
+ */
+static bool
+FilterShardsFromPgclass(Node *node, void *context)
+{
+	if (node == NULL)
 	{
-		return query_tree_walker((Query *) inputNode, ReplaceTableVisibleFunctionWalker,
-								 NULL, 0);
+		return false;
 	}
 
-	return expression_tree_walker(inputNode, ReplaceTableVisibleFunctionWalker, NULL);
+	if (IsA(node, Query))
+	{
+		Query *query = (Query *) node;
```

Comment:
I think no one should ever run non-select queries over pg_catalog tables.

Still, it feels slightly safer to kick in this logic only for SELECT-only queries? I'm mostly concerned with: `UPDATE non-pg-catalog-table FROM pg_class JOIN ON oid ..` type of queries.

We should probably let them happen irrespective of the app name?

Reply:
makes sense

Reply:
I played around with it, but it seems unintuitive that queries sometimes behave differently (e.g. `WITH event_tables as (SELECT relname FROM pg_class WHERE relname LIKE 'events%'), up AS (UPDATE ...) SELECT count(*) FROM up;` or `SELECT oid FROM pg_class WHERE relnamespace = 12345 ORDER BY relname LIMIT 1 FOR UPDATE; UPDATE pg_class SET  WHERE oid = $1;`  

I did verify that insert/update/delete works fine.